src/main/java/it/pagopa/transactions/mdcutilities/MDCFilter.java

```diff
@@ -0,0 +1,65 @@
+package it.pagopa.transactions.mdcutilities;
+
+import lombok.extern.slf4j.Slf4j;
+import org.slf4j.MDC;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.stereotype.Component;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.ServerWebExchangeDecorator;
+import org.springframework.web.server.WebFilter;
+import org.springframework.web.server.WebFilterChain;
+import org.springframework.web.util.UriTemplate;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Consumer;
+
+@Component
+@Slf4j
+public class MDCFilter implements WebFilter {
+
+    private ServerWebExchange decorate(ServerWebExchange exchange) {
+
+        final ServerHttpRequest decoratedRequest = new MDCCachingValuesServerHttpRequestDecorator(exchange.getRequest());
+
+        return new ServerWebExchangeDecorator(exchange) {
+
+            @Override
+            public ServerHttpRequest getRequest() {
+                return decoratedRequest;
+            }
+
+        };
+    }
+
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+        ServerHttpRequest request = exchange.getRequest();;
+        Map<String,String> transactionMap = getTransactionId(exchange.getRequest());
+        MDC.put("contextKey", getRequestId(request.getHeaders()));
+        Optional.ofNullable(transactionMap.get("transactionId")).ifPresent(v -> MDC.put("transactionId", v));
+        Optional.ofNullable(transactionMap.get("paymentContextCode")).ifPresent(v -> MDC.put("paymentContextCode", v));
+        return chain.filter(decorate(exchange));
+    }
+
+    private Map<String, String> getTransactionId(ServerHttpRequest request) {
+        UriTemplate uriTemplatePCC = new UriTemplate("/transactions/payment-context-codes/{paymentContextCode}/activation-results");
+        UriTemplate uriTemplateStandard = new UriTemplate("/transactions/{transactionId}");
+        return uriTemplatePCC.matches(request.getPath().value()) ? uriTemplatePCC.match(request.getPath().value()) : uriTemplateStandard.match(request.getPath().value());
+    }
+
+    private String getRequestId(HttpHeaders headers) {
+        //FIXME put here appInsight id in place of "X-Request-ID"
```

Comment:
like this https://learn.microsoft.com/en-us/azure/azure-monitor/app/correlation#correlation-headers-using-w3c-tracecontext